File: scripts/import-crsp2025.js
#!/usr/bin/env node

import fs from 'fs';
import { parse } from 'csv-parse';
import { Pool } from '@neondatabase/serverless';

// Clean CRSP value string and convert to decimal
function cleanCrspValue(valueStr) {
  if (!valueStr || valueStr.trim() === '') {
    return null;
  }
  
  // Remove quotes, spaces, and commas
  const cleaned = valueStr.toString().trim().replace(/['"]+/g, '').replace(/,/g, '').replace(/ /g, '');
  
  // Try to extract numeric value
  try {
    return parseFloat(cleaned);
  } catch (error) {
    return null;
  }
}

// Extract numeric engine capacity from string
function cleanEngineCapacity(engineStr) {
  if (!engineStr || engineStr.trim() === '') {
    return null;
  }
  
  const str = engineStr.toString().trim();
  
  // Handle electric vehicles with kWh
  if (str.toUpperCase().includes('KWH')) {
    return null; // Skip electric vehicles for engine capacity
  }
  
  // Extract numeric values
  const numbers = str.match(/\d+/);
  if (numbers && numbers.length > 0) {
    return parseInt(numbers[0]);
  }
  
  return null;
}

// Clean seating number
function cleanSeating(seatingStr) {
  if (!seatingStr || seatingStr.trim() === '') {
    return null;
  }
  
  try {
    return parseInt(parseFloat(seatingStr)).toString();
  } catch (error) {
    return null;
  }
}

// Normalize fuel type values
function normalizeFuelType(fuelStr) {
  if (!fuelStr || fuelStr.trim() === '') {
    return null;
  }
  
  const fuelMap = {
    'GASOLINE': 'petrol',
    'PETROL': 'petrol',
    'DIESEL': 'diesel',
    'ELECTRIC': 'electric',
    'HYBRID': 'hybrid',
    'PLUG-IN HYBRID': 'hybrid'
  };
  
  const fuelUpper = fuelStr.toString().trim().toUpperCase();
  return fuelMap[fuelUpper] || fuelUpper.toLowerCase();
}

async function main() {
  try {
    // Database connection
    const DATABASE_URL = process.env.DATABASE_URL;
    if (!DATABASE_URL) {
      console.error('ERROR: DATABASE_URL environment variable not set');
      return;
    }
    
    const pool = new Pool({ connectionString: DATABASE_URL });
    
    // Read and parse CSV file
    const csvContent = fs.readFileSync('attached_assets/CRSP2025_1753595396478.csv', 'utf-8');
    
    const parser = parse(csvContent, {
      columns: true,
      skip_empty_lines: true,
      trim: true
    });
    
    const rows = [];
    for await (const row of parser) {
      rows.push(row);
    }
    
    console.log(`Found ${rows.length} records in CRSP 2025 CSV`);
    
    let successfulImports = 0;
    let skippedRecords = 0;
    
    for (let index = 0; index < rows.length; index++) {
      const row = rows[index];
      
      try {
        const make = row['Make'] && row['Make'].trim() ? row['Make'].trim().toUpperCase() : null;
        const model = row['Model'] && row['Model'].trim() ? row['Model'].trim() : null;
        const modelNumber = row['Model \nnumber'] && row['Model \nnumber'].trim() ? row['Model \nnumber'].trim() : null;
        const transmission = row['Transmission'] && row['Transmission'].trim() ? row['Transmission'].trim() : null;
        const driveConfig = row['Drive\nConfiguration'] && row['Drive\nConfiguration'].trim() ? row['Drive\nConfiguration'].trim() : null;
        const engineCapacity = cleanEngineCapacity(row['Engine \nCapacity']);
        const bodyType = row['Body \nType '] && row['Body \nType '].trim() ? row['Body \nType '].trim() : null;
        const gvw = row['GVW'] && row['GVW'].trim() ? row['GVW'].trim() : null;
        const seating = cleanSeating(row['Seating']);
        const fuelType = normalizeFuelType(row['Fuel']);
        const crsp2025 = cleanCrspValue(row['CRSP2025']);
        
        if (!make || !model || !crsp2025) {
          console.log(`Skipping row ${index}: Missing required data (make=${make}, model=${model}, crsp2025=${crsp2025})`);
          skippedRecords++;
          continue;
        }
        
        // Check if record already exists (by make, model, and model_number)
        const existingQuery = `
          SELECT id FROM vehicle_references 
          WHERE UPPER(make) = $1 AND model = $2 AND ($3::text IS NULL OR model_number = $3)
        `;
        const existingResult = await pool.query(existingQuery, [make, model, modelNumber]);
        
        if (existingResult.rows.length > 0) {
          // Update existing record with CRSP2025 value
          const updateQuery = `
            UPDATE vehicle_references 
            SET crsp_2025 = $1,
                transmission = $2,
                drive_configuration = $3,
                engine_capacity = $4,
                body_type = $5,
                gvw = $6,
                seating = $7,
                fuel_type = $8,
                model_number = $9
            WHERE id = $10
          `;
          await pool.query(updateQuery, [
            crsp2025, transmission, driveConfig, engineCapacity,
            bodyType, gvw, seating, fuelType, modelNumber, existingResult.rows[0].id
          ]);
        } else {
          // Insert new record
          const insertQuery = `
            INSERT INTO vehicle_references 
            (make, model, model_number, transmission, drive_configuration, 
             engine_capacity, body_type, gvw, seating, fuel_type, crsp_2025)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
          `;
          await pool.query(insertQuery, [
            make, model, modelNumber, transmission, driveConfig,
            engineCapacity, bodyType, gvw, seating, fuelType, crsp2025
          ]);
        }
        
        successfulImports++;
        
        if (successfulImports % 100 === 0) {
          console.log(`Processed ${successfulImports} records...`);
        }
        
      } catch (error) {
        console.error(`Error processing row ${index}:`, error.message);
        continue;
      }
    }
    
    console.log(`\nImport completed:`);
    console.log(`Successfully imported/updated: ${successfulImports} records`);
    console.log(`Skipped records: ${skippedRecords}`);
    
    // Verify import
    const countResult = await pool.query('SELECT COUNT(*) FROM vehicle_references WHERE crsp_2025 IS NOT NULL');
    console.log(`Total records with CRSP 2025 values: ${countResult.rows[0].count}`);
    
    await pool.end();
    
  } catch (error) {
    console.error('Import failed:', error);
  }
}

main();

File: scripts/import-crsp2025.py
#!/usr/bin/env python3
"""
Script to import CRSP 2025 data into the vehicle_references table
"""

import csv
import psycopg2
import os
import re

def clean_crsp_value(value_str):
    """Clean CRSP value string and convert to decimal"""
    if not value_str or value_str.strip() == '':
        return None
    
    # Remove quotes, spaces, and commas
    cleaned = str(value_str).strip().replace('"', '').replace("'", "").replace(',', '').replace(' ', '')
    
    # Try to extract numeric value
    try:
        return float(cleaned)
    except ValueError:
        return None

def clean_engine_capacity(engine_str):
    """Extract numeric engine capacity from string"""
    if not engine_str or engine_str.strip() == '':
        return None
    
    engine_str = str(engine_str).strip()
    
    # Handle electric vehicles with kWh
    if 'kWh' in engine_str.upper():
        return None  # Skip electric vehicles for engine capacity
    
    # Extract numeric values
    numbers = re.findall(r'\d+', engine_str)
    if numbers:
        return int(numbers[0])
    
    return None

def clean_seating(seating_str):
    """Clean seating number"""
    if not seating_str or seating_str.strip() == '':
        return None
    
    try:
        return str(int(float(seating_str)))
    except (ValueError, TypeError):
        return None

def normalize_fuel_type(fuel_str):
    """Normalize fuel type values"""
    if not fuel_str or fuel_str.strip() == '':
        return None
    
    fuel_map = {
        'GASOLINE': 'petrol',
        'PETROL': 'petrol', 
        'DIESEL': 'diesel',
        'ELECTRIC': 'electric',
        'HYBRID': 'hybrid',
        'PLUG-IN HYBRID': 'hybrid'
    }
    
    fuel_upper = str(fuel_str).strip().upper()
    return fuel_map.get(fuel_upper, fuel_upper.lower())

def main():
    # Database connection
    DATABASE_URL = os.getenv('DATABASE_URL')
    if not DATABASE_URL:
        print("ERROR: DATABASE_URL environment variable not set")
        return
    
    conn = psycopg2.connect(DATABASE_URL)
    cur = conn.cursor()
    
    # Read CSV file
    with open('attached_assets/CRSP2025_1753595396478.csv', 'r', encoding='utf-8') as csvfile:
        # Read first to get total count
        reader = csv.DictReader(csvfile)
        rows = list(reader)
    
    print(f"Found {len(rows)} records in CRSP 2025 CSV")
    
    # Process each row
    successful_imports = 0
    skipped_records = 0
    
    for index, row in enumerate(rows):
        try:
            make = str(row['Make']).strip().upper() if row['Make'] and row['Make'].strip() else None
            model = str(row['Model']).strip() if row['Model'] and row['Model'].strip() else None
            model_number = str(row['Model \nnumber']).strip() if row['Model \nnumber'] and row['Model \nnumber'].strip() else None
            transmission = str(row['Transmission']).strip() if row['Transmission'] and row['Transmission'].strip() else None
            drive_config = str(row['Drive\nConfiguration']).strip() if row['Drive\nConfiguration'] and row['Drive\nConfiguration'].strip() else None
            engine_capacity = clean_engine_capacity(row['Engine \nCapacity'])
            body_type = str(row['Body \nType ']).strip() if row['Body \nType '] and row['Body \nType '].strip() else None
            gvw = str(row['GVW']).strip() if row['GVW'] and row['GVW'].strip() else None
            seating = clean_seating(row['Seating'])
            fuel_type = normalize_fuel_type(row['Fuel'])
            crsp2025 = clean_crsp_value(row['CRSP2025'])
            
            if not make or not model or not crsp2025:
                print(f"Skipping row {index}: Missing required data (make={make}, model={model}, crsp2025={crsp2025})")
                skipped_records += 1
                continue
            
            # Check if record already exists (by make, model, and model_number)
            cur.execute("""
                SELECT id FROM vehicle_references 
                WHERE UPPER(make) = %s AND model = %s AND model_number = %s
            """, (make, model, model_number))
            
            existing = cur.fetchone()
            
            if existing:
                # Update existing record with CRSP2025 value
                cur.execute("""
                    UPDATE vehicle_references 
                    SET crsp_2025 = %s,
                        transmission = %s,
                        drive_configuration = %s,
                        engine_capacity = %s,
                        body_type = %s,
                        gvw = %s,
                        seating = %s,
                        fuel_type = %s
                    WHERE id = %s
                """, (crsp2025, transmission, drive_config, engine_capacity, 
                     body_type, gvw, seating, fuel_type, existing[0]))
            else:
                # Insert new record
                cur.execute("""
                    INSERT INTO vehicle_references 
                    (make, model, model_number, transmission, drive_configuration, 
                     engine_capacity, body_type, gvw, seating, fuel_type, crsp_2025)
                    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                """, (make, model, model_number, transmission, drive_config,
                     engine_capacity, body_type, gvw, seating, fuel_type, crsp2025))
            
            successful_imports += 1
            
            if successful_imports % 100 == 0:
                print(f"Processed {successful_imports} records...")
                conn.commit()
                
        except Exception as e:
            print(f"Error processing row {index}: {e}")
            continue
    
    # Final commit
    conn.commit()
    
    print(f"\nImport completed:")
    print(f"Successfully imported/updated: {successful_imports} records")
    print(f"Skipped records: {skipped_records}")
    
    # Verify import
    cur.execute("SELECT COUNT(*) FROM vehicle_references WHERE crsp_2025 IS NOT NULL")
    count = cur.fetchone()[0]
    print(f"Total records with CRSP 2025 values: {count}")
    
    cur.close()
    conn.close()

if __name__ == "__main__":
    main()